components/ac_log/src/lib.rs

```diff
@@ -0,0 +1,280 @@
+//!    somewhere else happens to log something), we use a separate channel that
+//!    only exists to indicate that the `Arc<AtomicBool>` has changed value.
+//!
+//! (For future work it might be worth investigate if we can avoid the channel
+//! in `2.` by just closing (e.g. dropping) the log message channel's Sender and
+//! handling the Err in the background thread. This seems like it would work,
+//! but it's possible there are subtle issues).
+//!
+//! Finally, it's worth noting that the log crate is rather inflexable, in that
+//! it does not allow users to change loggers after the first initialization.
+//! This inflexibility has leaked into this API, but it's a consequence of
+//! `log`'s design, and not of anything fundamental about code calling between
+//! Rust and the JVM in this manner.
+
+use crossbeam_channel::Sender;
+use std::{
+    ffi::CString,
+    os::raw::c_char,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread,
+};
+
+#[derive(Clone, Copy)]
+#[repr(i32)]
+pub enum LogLevel {
+    // Android logger levels
+    VERBOSE = 2,
+    DEBUG = 3,
+    INFO = 4,
+    WARN = 5,
+    ERROR = 6,
+}
+
+impl From<log::Level> for LogLevel {
+    fn from(l: log::Level) -> Self {
+        match l {
+            log::Level::Trace => LogLevel::VERBOSE,
+            log::Level::Debug => LogLevel::DEBUG,
+            log::Level::Info => LogLevel::INFO,
+            log::Level::Warn => LogLevel::WARN,
+            log::Level::Error => LogLevel::ERROR,
+        }
+    }
+}
+
+// TODO: use serde to send this to the other thread as bincode or something,
+// rather than allocating all these strings for every message.
+struct LogRecord {
+    level: LogLevel,
+    tag: Option<CString>,
+    message: CString,
+}
+
+fn string_to_cstring_lossy(s: String) -> CString {
+    let mut bytes = s.into_bytes();
+    for byte in bytes.iter_mut() {
+        if *byte == 0 {
+            *byte = b'?';
+        }
+    }
+    CString::new(bytes).expect("Bug in string_to_cstring_lossy!")
+}
+
+impl<'a, 'b> From<&'b log::Record<'a>> for LogRecord {
+    // XXX important! Don't log in this function!
+    fn from(r: &'b log::Record<'a>) -> Self {
+        let message = match (r.line(), r.file()) {
+            (Some(line), Some(file)) => format!("{}@{}: {}", file, line, r.args()),
+            (None, Some(file)) => format!("{}: {}", file, r.args()),
+            // The (Some(line), None) case is pointless
+            (_, None) => format!("{}", r.args()),
+        };
+
+        Self {
+            level: r.level().into(),
+            tag: r
+                .module_path()
+                .and_then(|mp| CString::new(mp.to_owned()).ok()),
+            message: string_to_cstring_lossy(message),
+        }
+    }
+}
+
+/// Type of the log callback provided to us by java.
+/// Takes the following arguments:
+///
+/// - Log level (an i32).
+/// - Tag: a (nullable) nul terminated c string. Caller must not free this string!
```

Comment:
s/nul/null/

Reply:
AFAIK, NUL is the correct way too to refer to the 0 byte, as awkward as it is. Rust uses it too (for example, https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_bytes_with_nul), as does e.g. man ascii.

Wikipedia seems to say they're both right, which is news to me. It doesn't matter much though, and does look weird, so I'll change it if you want.

Reply:
I do not care strongly.  I've seen it but it's less usual.  If the Rust community likes it, that's good enough for me!